File: 8_string_to_integer/go/main.go
package main

import (
	"fmt"
)

func myAtoi(str string) int {
	base, sign, i := 0, 1, 0
	for ; i < len(str) && str[i] == ' '; i++ {}
	if i < len(str) && (str[i] == '-' || str[i] == '+') {
		if (str[i] == '-') {
			sign = -1
		}
		i++
	}
	for ; i < len(str) && str[i] >= '0' && str[i] <= '9'; i++ {
		base = base * 10 + int(str[i] - '0')
		if base > 2147483647 || base < -2147483648 {
			if sign == 1 {
				return 2147483647
			}
			return -2147483648
		}
	}
	return base * sign
}

func main() {
	var s string

	s = "42"
	fmt.Println(myAtoi(s))

	s = "   -42"
	fmt.Println(myAtoi(s))

	s = "4193 with words"
	fmt.Println(myAtoi(s))

	s = "words and 987"
	fmt.Println(myAtoi(s))

	s = "-91283472332"
	fmt.Println(myAtoi(s))
}

File: 8_string_to_integer/main.py
# -*- coding: utf-8 -*-


class Solution(object):

    def my_atoi(self, s):
        if s == '':
            return 0;
        base, sign, i = 0, 1, 0
        while s[i] == ' ':
            i += 1
        if i < len(s) and s[i] in ('-', '+'):
            sign = 1 - 2 * (s[i] == '-')
            i += 1

        while i < len(s) and s[i] >= '0' and s[i] <= '9':
            base = base * 10 + int(s[i])
            if base > 2 ** 31 - 1 or base < -(2 ** 31):
                return 2 ** 31 - 1 if sign == 1 else -(2 ** 31)
            i += 1

        return base * sign


if __name__ == '__main__':
    sl = Solution()

    s = '42'
    print(sl.my_atoi(s))

    s = '   -42'
    print(sl.my_atoi(s))

    s = '4193 with words'
    print(sl.my_atoi(s))

    s = 'words and 987'
    print(sl.my_atoi(s))

    s = '-91283472332'
    print(sl.my_atoi(s))
